### PR TITLE
Added design for FSM

### PR DESCRIPTION
**Pull request by BtXin**
**What this PR does / why we need it**:
Design for FSM feature
**Will this PR make the community happier**? 
Yes
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**How is this PR tested**
- [ ] unit test
- [ ] e2e test
- [x] other (please specify)
Docs

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: docs/design/design-shifuFSM.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
+# Shifu FSM Design
+Shifu FSM stands for Shifu Finite State Machine. Shifu FSM stands for Shifu Finite State Machine. It is a workflow control system that allows users to control the behavior of the device based on different states.
```
Comment (kris21he) [inline docs/design/design-shifuFSM.md:2]: duplicate description? "Shifu FSM stands for Shifu Finite State Machine."

File: docs/design/design-shifuFSM.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+## Control Flow
```
Comment (kris21he) [inline docs/design/design-shifuFSM.md:4]: make a real example of FSM seems make more sense

File: docs/design/design-shifuFSM.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+stateA -->|action1| stateB
+stateA -->|action2| stateC
+stateC -->|action3| stateA
+
+```
+
+### State
+A state is a condition or status of the system. It represents a specific moment or situation in the execution of the machine.
+
+### Action
```
Comment (kris21he) [inline docs/design/design-shifuFSM.md:16]: missed "forbid"?

File: docs/design/design-shifuFSM.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+An action here is just an instruction in the configmap. Each state can define a set of instructions along with the next state the device will end up in when receiving the action. If a state receives an unsupported action, it will simply rejects the action.
+
+Take the above control flow as an example, `stateA` receives `action1` and transit to `stateB`. `stateA` receives `action2` and transit to `stateC`. `stateC` receives `action3` and transit back to `stateA`.
+
+
+## High level design
+We can reuse a lot of existing design and implementations to achieve the FSM functionality. We can embed the entire FSM as part of our configmap, along with instructions and telemetries.
```
Comment (kris21he) [inline docs/design/design-shifuFSM.md:23]: "along with instructions and telemetries." seems need to add telemetries design? only see how it embed with instructions. if we don't want to support telemetry this time, we can remove it?

File: docs/design/design-shifuFSM.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+Take the above control flow as an example, `stateA` receives `action1` and transit to `stateB`. `stateA` receives `action2` and transit to `stateC`. `stateC` receives `action3` and transit back to `stateA`.
+
+
+## High level design
+We can reuse a lot of existing design and implementations to achieve the FSM functionality. We can embed the entire FSM as part of our configmap, along with instructions and telemetries. 
+We can add a state field inside `edgedeivce` to represent current state of the device, and since edgedevice is stored in `etcd`, it's also good for fail-recovery.
+
+## Detailed Design
+### Configmap
```
Comment (kris21he) [inline docs/design/design-shifuFSM.md:27]: also put an example of configmap yaml path in shifu repository here?

File: docs/design/design-shifuFSM.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+            next_state: yellow
+        forbid:
+          stop:
+      yellow:
+        actions:
+          go:
+            next_state: green
+          stop:
+            next_state: red     
+    startingState: red
```
Comment (kris21he) [inline docs/design/design-shifuFSM.md:63]: startingState: simplify to -> init:

File: docs/design/design-shifuFSM.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+}
+```
+We will add a new filed called `State` in `EdgeDeviceSpec` to represent the current state of the device.
+
+We will handle the FSM transition in `commandHandleFunc` inside each deviceShifu subtype/
+```golang
+func (handler DeviceCommandHandlerHTTP) commandHandleFunc() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+        // Get associated instruction and current state
+        state := fsm.States[state]
```
Comment (kris21he) [inline docs/design/design-shifuFSM.md:150]: check if no FSM set, we'd ignore

File: docs/design/design-shifuFSM.md [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+```golang
+// DeviceShifuConfig data under Configmap, Settings of deviceShifu
+type DeviceShifuConfig struct {
+	DriverProperties         DeviceShifuDriverProperties
+	Instructions             DeviceShifuInstructions
+	Telemetries              *DeviceShifuTelemetries
+	CustomInstructionsPython map[string]string `yaml:"customInstructionsPython"`
+	ControlMsgs              map[string]string `yaml:"controlMsgs,omitempty"`
+	FSM                      DeviceShifuFSM
```
Comment (kris21he) [inline docs/design/design-shifuFSM.md:108]: *DeviceShifuFSM?

File: docs/design/design-shifuFSM.md [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+type DeviceShifuFSM struct {
+	States        map[string]*DeviceShifuState `yaml:"states"`
+	StartingState string                       `yaml:"startingState"`
```
Comment (kris21he) [inline docs/design/design-shifuFSM.md:113]: *string?

File: docs/design/design-shifuFSM.md [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+  name: http-configmap-0.0.1
+  namespace: deviceshifu
+data:
+#    available instructions
+  instructions: |
+    instructions:
+      caution:
+      stop:
+      go:
+  fsm: |
```
Comment (kris21he) [inline docs/design/design-shifuFSM.md:43]: do we need a design to check if it is a valid fsm config?

File: docs/design/design-shifuFSM.md (hunk 4 above)
Comment (BtXin) [reply docs/design/design-shifuFSM.md:23]: Telemetry is a existing feature sit alongside with instructions. I will added telemetry in the configmap as well.

File: docs/design/design-shifuFSM.md (hunk 5 above)
Comment (BtXin) [reply docs/design/design-shifuFSM.md:27]: The configmap here is already a valid yaml file, don't think it is necessary to put another example.

File: docs/design/design-shifuFSM.md (hunk 6 above)
Comment (BtXin) [reply docs/design/design-shifuFSM.md:63]: init seems too vague and oversimplified, will keep startingState here.

File: docs/design/design-shifuFSM.md (hunk 10 above)
Comment (BtXin) [reply docs/design/design-shifuFSM.md:43]: That can be a separate task if we do need to have a check.

**Review by leepala — DISMISSED** [review]

File: docs/design/design-shifuFSM.md [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+### Action
+An action here is just an instruction in the configmap. Each state can define a set of instructions along with the next state the device will end up in when receiving the action. If a state receives an unsupported action, it will simply rejects the action.
+
+Take the above control flow as an example, `green` receives `caution` and transit to `yellow`. `yellow` receives `stop` and transit to `red`. `red` receives `caution` and transit back to `yellow`, `yellow` receives `go` action and transit to `green`.
+
+### Forbid
+Forbid will be a list of actions that are not allowed to take in the given state. In the traffic light example, a `red` state would not allowed to take a `go` action directly.
+
+
+## High level design
```
Comment (kris21he) [inline docs/design/design-shifuFSM.md:27]: suggest move the control flow under High level design?

File: docs/design/design-shifuFSM.md [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+    Dimensions: 956*730*240 mm
+    RatedLoad: Rated load
+    BatteryCapacity: 48V/36Ah
+    ChargingTime: 9h
+    AmbientTemperature: 0-45C
+  state: green
+```
+We will add a new filed `state` in `edgeDevice` spec to represent current state of the device.
+
+### Code
```
Comment (kris21he) [inline docs/design/design-shifuFSM.md:111]: Psuedo-code?

**Review by kris21he — APPROVED** [review]
LGTM

**Pull request merged**